File: 2023/rust/src/bin/d07p1.rs
use std::{collections::HashMap, fs};

#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
enum Card {
    Two,
    Three,
    Four,
    Five,
    Six,
    Seven,
    Eight,
    Nine,
    Ten,
    Jack,
    Queen,
    King,
    Ace,
}
impl From<char> for Card {
    fn from(value: char) -> Self {
        match value {
            '2' => Card::Two,
            '3' => Card::Three,
            '4' => Card::Four,
            '5' => Card::Five,
            '6' => Card::Six,
            '7' => Card::Seven,
            '8' => Card::Eight,
            '9' => Card::Nine,
            'T' => Card::Ten,
            'J' => Card::Jack,
            'Q' => Card::Queen,
            'K' => Card::King,
            'A' => Card::Ace,
            _ => panic!("Unknown card: {value}"),
        }
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
enum HandType {
    HighCard,
    OnePair,
    TwoPair,
    ThreeKind,
    FullHouse,
    FourKind,
    FiveKind,
}
impl From<[Card; 5]> for HandType {
    fn from(value: [Card; 5]) -> Self {
        let mut map: HashMap<Card, usize> = HashMap::new();
        for c in value {
            *map.entry(c).or_default() += 1;
        }

        let length = map.keys().len();
        match length {
            1 => HandType::FiveKind,
            2 => {
                let vals: Vec<&usize> = map.values().collect();
                if *vals[0] == 4 || *vals[1] == 4 {
                    HandType::FourKind
                } else {
                    HandType::FullHouse
                }
            }
            3 => {
                let vals: Vec<&usize> = map.values().collect();
                if *vals[0] == 3 || *vals[1] == 3 || *vals[2] == 3 {
                    HandType::ThreeKind
                } else {
                    HandType::TwoPair
                }
            }
            4 => HandType::OnePair,
            5 => HandType::HighCard,
            _ => panic!("Unknown hand type"),
        }
    }
}

#[derive(Debug, PartialEq, Eq)]
struct Hand {
    cards: [Card; 5],
    hand_type: HandType,
    bet: u32,
}
impl PartialOrd for Hand {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}
impl Ord for Hand {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        if !self.hand_type.eq(&other.hand_type) {
            return self.hand_type.cmp(&other.hand_type);
        } else {
            for (s, o) in self.cards.iter().zip(other.cards.iter()) {
                if !s.eq(o) {
                    return s.cmp(o);
                }
            }
        }

        std::cmp::Ordering::Equal
    }
}

fn main() {
    println!("Day 7 - Part 1");

    let str = fs::read_to_string("../_input/day07.txt").unwrap();
    let lines: Vec<&str> = str.lines().collect();

    let total = process(lines);
    println!("Total: {}", total);
}

fn process(lines: Vec<&str>) -> u32 {
    let mut hands: Vec<Hand> = lines.iter().map(|i| process_line(i)).collect();
    hands.sort();

    let mut counter = 0;
    for (i, h) in hands.iter().enumerate() {
        counter += h.bet * (i as u32 + 1);
    }

    counter
}

fn process_line(line: &str) -> Hand {
    let s: Vec<&str> = line.split_whitespace().collect();

    let mut cards: [Card; 5] = [Card::Two; 5];
    for (i, c) in s[0].chars().enumerate() {
        cards[i] = c.into();
    }

    Hand {
        cards,
        hand_type: cards.into(),
        bet: str_to_u32(s[1]),
    }
}

fn str_to_u32(str: &str) -> u32 {
    let mut acc = 0;
    for (i, c) in str.chars().enumerate() {
        acc += (10_u32.pow((str.len() - i - 1) as u32)) * c.to_digit(10).unwrap();
    }
    acc
}

#[cfg(test)]
mod test {
    use crate::process;

    #[test]
    fn test() {
        assert_eq!(
            process(
                r"32T3K 765
T55J5 684
KK677 28
KTJJT 220
QQQJA 483"
                    .split('\n')
                    .collect()
            ),
            6440
        );

        assert_eq!(
            process(
                r"2345A 1
Q2KJJ 13
Q2Q2Q 19
T3T3J 17
T3Q33 11
2345J 3
J345A 2
32T3K 5
T55J5 29
KK677 7
KTJJT 34
QQQJA 31
JJJJJ 37
JAAAA 43
AAAAJ 59
AAAAA 61
2AAAA 23
2JJJJ 53
JJJJ2 41"
                    .split('\n')
                    .collect()
            ),
            6592
        );
    }
}

File: 2023/java/d07p1.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;

public class d07p1 {
    private static final int ZERO = 48;

    public static void main(String[] args) throws IOException {
        System.out.println("Day 7 - Part 1");

        BufferedReader reader = new BufferedReader(new FileReader("../_input/day07.txt"));

        long total = process(reader);
        reader.close();

        System.out.println("Total: " + total);
    }

    private static int process(BufferedReader reader) throws IOException {
        ArrayList<Hand> hands = new ArrayList<>();
        String line;
        while ((line = reader.readLine()) != null) hands.add(processLine(line));

        Collections.sort(hands);

        int counter = 0;
        for (int i = 0; i < hands.size(); i++) counter += hands.get(i).bet * (i + 1);
        return counter;
    }

    private static Hand processLine(String line) {
        String[] s = line.split(" ");

        Card[] cards = {Card.TWO, Card.TWO, Card.TWO, Card.TWO, Card.TWO};
        int[] cs = s[0].chars().toArray();
        for (int i = 0; i < cs.length; i++) cards[i] = Card.get(cs[i]);

        return new Hand(cards, HandType.get(cards), stringToInt(s[1]));
    }

    private static int stringToInt(String str) {
        int acc = 0;
        for (int i = 0; i < str.length(); i++) {
            int c = str.charAt(i);
            acc += (int) (Math.pow(10.0, str.length() - i - 1) * (c - ZERO));
        }
        return acc;
    }

    private enum Card {
        TWO(0),
        THREE(1),
        FOUR(2),
        FIVE(3),
        SIX(4),
        SEVEN(5),
        EIGHT(7),
        NINE(8),
        TEN(9),
        JACK(10),
        QUEEN(11),
        KING(12),
        ACE(13);

        public final int label;

        Card(int label) {
            this.label = label;
        }

        public static Card get(int c) {
            switch (c) {
                case (int) '2' -> {
                    return Card.TWO;
                }
                case (int) '3' -> {
                    return Card.THREE;
                }
                case (int) '4' -> {
                    return Card.FOUR;
                }
                case (int) '5' -> {
                    return Card.FIVE;
                }
                case (int) '6' -> {
                    return Card.SIX;
                }
                case (int) '7' -> {
                    return Card.SEVEN;
                }
                case (int) '8' -> {
                    return Card.EIGHT;
                }
                case (int) '9' -> {
                    return Card.NINE;
                }
                case (int) 'T' -> {
                    return Card.TEN;
                }
                case (int) 'J' -> {
                    return Card.JACK;
                }
                case (int) 'Q' -> {
                    return Card.QUEEN;
                }
                case (int) 'K' -> {
                    return Card.KING;
                }
                case (int) 'A' -> {
                    return Card.ACE;
                }
                default -> throw new RuntimeException("Could not convert int to card.");
            }
        }
    }

    private enum HandType {
        HIGH_CARD(0),
        ONE_PAIR(1),
        TWO_PAIR(2),
        THREE_KIND(3),
        FULL_HOUSE(4),
        FOUR_KIND(5),
        FIVE_KIND(6);

        public final int label;

        HandType(int label) {
            this.label = label;
        }

        public static HandType get(Card[] cards) {
            if (cards.length != 5) throw new RuntimeException("Cards array is not length of five.");

            HashMap<Card, Integer> map = new HashMap<>();
            for (Card c : cards) {
                var i = map.getOrDefault(c, 0);
                map.put(c, ++i);
            }

            int length = map.keySet().size();
            switch (length) {
                case 1 -> {
                    return HandType.FIVE_KIND;
                }
                case 2 -> {
                    if (map.containsValue(4)) return HandType.FOUR_KIND;
                    else return HandType.FULL_HOUSE;
                }
                case 3 -> {
                    if (map.containsValue(3)) return HandType.THREE_KIND;
                    else return HandType.TWO_PAIR;
                }
                case 4 -> {
                    return HandType.ONE_PAIR;
                }
                case 5 -> {
                    return HandType.HIGH_CARD;
                }
                default -> throw new RuntimeException("Invalid hand.");
            }
        }
    }

    private record Hand(Card[] cards, HandType handType, int bet) implements Comparable<Hand> {
        @Override
        public int compareTo(Hand o) {
            if (handType != o.handType) return handType.label - o.handType.label;
            else {
                for (int i = 0; i < cards.length; i++) {
                    var a = cards[i];
                    var b = o.cards[i];
                    if (a != b) return a.label - b.label;
                }
            }
            return 0;
        }
    }
}
